resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerResponseFeedbackAddressStream.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2022 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.resolver.dns;
+
+import java.net.InetSocketAddress;
+
+/**
+ * An infinite stream of DNS server addresses, that requests feedback to be returned to it.
+ *
+ * If query is successful timing information is provided, else a failure notification is given.
+ */
+public interface DnsServerResponseFeedbackAddressStream extends DnsServerAddressStream {
+
+    /**
+     * A way to provide timing feedback to {@link DnsServerAddressStream} so that {@link #next()} can be tuned
+     * to return the best performing DNS server address
+     *
+     * NOTE: This is called regardless of the RCode returned by the DNS server
+     *
+     * @param address The address returned by {@link #next()} that feedback needs to be applied to
+     * @param queryResponseTimeNanos The response time of a query against the given DNS server
+     */
+    void feedbackSuccess(InetSocketAddress address, long queryResponseTimeNanos);
+
+    /**
+     * A way to provide failure feedback to {@link DnsServerAddressStream} so that {@link #next()} cab be tuned
+     * to return the best performing DNS server address
+     *
+     * @param address The address returned by {@link #next()} that feedback needs to be applied to
+     */
+    void feedbackFailure(InetSocketAddress address);
```

Comment:
@watson-ryan I wonder if we also want to include the nanos here ? This way we know if it's a timeout or something else . WDYT ?

Reply:
I don't see any harm in including it, implementations can ignore it if they don't want to try to distinguish the type of failure.

I do believe it will be difficult for users to definitively categorize the failure type with just the timing data, but it might be able to provide _something_

Reply:
yeah... we could also include the exception